Crash with IdeaVim 2.25.0-eap.3

After updating IdeaVim to 2.25.0-eap.3, WhichKey (0.10.3) will crash on any keypress and completely disable functionality, even just normal IdeaVim editing, leading to need for you to disable the plugin. 

This seems to be cause of using a deprecated API (com/maddyhome/idea/vim/command/CommandState.kt), which was recently removed <https://github.com/JetBrains/ideavim/commit/ab635972cc6cab7275658c545674262d35819e27>  

It seems like the alternative is to use `injector.vimState` (according to the deprecated marker)

<details>
<summary>Full stacktrace</summary>

```java
java.lang.NoClassDefFoundError: com/maddyhome/idea/vim/command/CommandState
	at eu.theblob42.idea.whichkey.WhichKeyTypeActionHandler.beforeExecute(WhichKeyTypeActionHandler.kt:39)
	at com.intellij.openapi.editor.actionSystem.TypedAction.beforeActionPerformed(TypedAction.java:191)
	at com.intellij.openapi.editor.impl.EditorImpl.processKeyTypedImmediately(EditorImpl.java:1449)
	at com.intellij.openapi.editor.impl.EditorImpl.processKeyTyped(EditorImpl.java:1434)
	at com.intellij.openapi.editor.impl.EditorImpl.processKeyTyped(EditorImpl.java:3837)
	at com.intellij.openapi.editor.impl.EditorImpl$7.keyTyped(EditorImpl.java:1272)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.AWTEventMulticaster.keyTyped(AWTEventMulticaster.java:247)
	at java.desktop/java.awt.Component.processKeyEvent(Component.java:6628)
	at java.desktop/javax.swing.JComponent.processKeyEvent(JComponent.java:2901)
	at java.desktop/java.awt.Component.processEvent(Component.java:6450)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5043)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4871)
	at java.desktop/java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:886)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1162)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:1023)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:849)
	at com.intellij.ide.IdeKeyboardFocusManager.dispatchEvent$lambda$1(IdeKeyboardFocusManager.kt:48)
	at com.intellij.ide.IdeKeyboardFocusManager.dispatchEvent$lambda$2(IdeKeyboardFocusManager.kt:54)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$3(IdeEventQueue.kt:958)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:109)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:958)
	at com.intellij.ide.IdeKeyboardFocusManager.dispatchEvent(IdeKeyboardFocusManager.kt:54)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4920)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2810)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4871)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:783)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:728)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:98)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:755)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:753)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:752)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:585)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.kt:518)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$17(IdeEventQueue.kt:474)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteIntentReadAction$lambda$6(AnyThreadWriteThreadingSupport.kt:274)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWithTemporaryThreadLocal(AnyThreadWriteThreadingSupport.kt:204)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runWriteIntentReadAction(AnyThreadWriteThreadingSupport.kt:274)
	at com.intellij.openapi.application.impl.AnyThreadWriteThreadingSupport.runPreventiveWriteIntentReadAction(AnyThreadWriteThreadingSupport.kt:218)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:474)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$12$lambda$11$lambda$10$lambda$9(IdeEventQueue.kt:307)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:864)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$12$lambda$11$lambda$10(IdeEventQueue.kt:306)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$3(IdeEventQueue.kt:958)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:117)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:958)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$12(IdeEventQueue.kt:301)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:341)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.lang.ClassNotFoundException: com.maddyhome.idea.vim.command.CommandState PluginClassLoader(plugin=PluginDescriptor(name=Which-Key, id=eu.theblob42.idea.whichkey, descriptorPath=plugin.xml, path=~/.local/share/JetBrains/IntelliJIdea2025.1/IDEA_Which-Key-0.10.3.jar, version=0.10.3, package=null, isBundled=false), packagePrefix=null, state=active, parents=PluginDescriptor(name=IDEA CORE, id=com.intellij, moduleName=intellij.platform.vcs.impl, descriptorPath=intellij.platform.vcs.impl.xml, path=~/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate/lib, version=251.23774.435, package=null, isBundled=true), PluginDescriptor(name=IDEA CORE, id=com.intellij, moduleName=intellij.platform.vcs.log.impl, descriptorPath=intellij.platform.vcs.log.impl.xml, path=~/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate/lib, version=251.23774.435, package=null, isBundled=true), PluginDescriptor(name=IDEA CORE, id=com.intellij, moduleName=intellij.platform.vcs.dvcs.impl, descriptorPath=intellij.platform.vcs.dvcs.impl.xml, path=~/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate/lib, version=251.23774.435, package=null, isBundled=true), PluginDescriptor(name=IDEA CORE, id=com.intellij, moduleName=intellij.platform.collaborationTools, descriptorPath=intellij.platform.collaborationTools.xml, path=~/.local/share/JetBrains/Toolbox/apps/intellij-idea-ultimate/lib, version=251.23774.435, package=null, isBundled=true), PluginDescriptor(name=IdeaVim, id=IdeaVIM, descriptorPath=plugin.xml, path=~/.local/share/JetBrains/IntelliJIdea2025.1/IdeaVIM, version=2.25.0-eap.3, package=null, isBundled=false), )
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClass(PluginClassLoader.kt:157)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 69 more
```
</details>

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -24,7 +24,7 @@ dependencies {
     intellijPlatform {
         intellijIdeaCommunity("2023.3.3")
         pluginVerifier()
-        plugins("IdeaVIM:2.10.0")
+        plugins("IdeaVIM:2.24.0")
     }
 }
 

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,7 +1,8 @@
+import org.jetbrains.kotlin.gradle.dsl.JvmTarget
 
 plugins {
     java
-    kotlin("jvm") version "1.9.22"
+    kotlin("jvm") version "2.1.21"
     id("org.jetbrains.intellij.platform") version "2.3.0"
 }
 
@@ -16,13 +17,10 @@ repositories {
 }
 
 dependencies {
-    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.4.21")
-    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:1.6.4")
-
     testImplementation("junit", "junit", "4.12")
 
     intellijPlatform {
-        intellijIdeaCommunity("2023.3.3")
+        intellijIdeaCommunity("2025.1")
         pluginVerifier()
         plugins("IdeaVIM:2.24.0")
     }
@@ -46,26 +44,26 @@ intellijPlatform {
 
 java {
     toolchain {
-        languageVersion.set(JavaLanguageVersion.of(17))
+        languageVersion.set(JavaLanguageVersion.of(21))
     }
 }
 
 kotlin {
     jvmToolchain {
-        languageVersion.set(JavaLanguageVersion.of(17))
+        languageVersion.set(JavaLanguageVersion.of(21))
     }
 }
 
 tasks {
     compileKotlin {
-        kotlinOptions {
-            jvmTarget = "17"
+        compilerOptions {
+            jvmTarget.set(JvmTarget.JVM_21)
         }
     }
 
     compileTestKotlin {
-        kotlinOptions {
-            jvmTarget = "17"
+        compilerOptions {
+            jvmTarget.set(JvmTarget.JVM_21)
         }
     }
 }

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/WhichKeyActionListener.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/WhichKeyActionListener.kt
@@ -7,11 +7,10 @@ import com.intellij.openapi.editor.actionSystem.ActionPlan
 import com.intellij.openapi.editor.actionSystem.TypedAction
 import com.intellij.openapi.editor.actionSystem.TypedActionHandlerEx
 import com.intellij.openapi.wm.WindowManager
-import com.maddyhome.idea.vim.KeyHandler
 import com.maddyhome.idea.vim.VimTypedActionHandler
+import com.maddyhome.idea.vim.KeyHandler
 import com.maddyhome.idea.vim.api.injector
 import com.maddyhome.idea.vim.command.Argument
-import com.maddyhome.idea.vim.command.CommandState
 import com.maddyhome.idea.vim.command.MappingMode
 import com.maddyhome.idea.vim.command.MappingState
 import com.maddyhome.idea.vim.impl.state.toMappingMode
@@ -64,8 +63,9 @@ class WhichKeyActionListener : AnActionListener {
         }
 
         val editor = dataContext.getData(CommonDataKeys.EDITOR) ?: return
-        val commandState = CommandState.getInstance(editor)
-        val vimCurrentKeySequence = commandState.mappingState.keys.toList().ifEmpty { commandState.commandBuilder.keys.toList() }
+        val keyHandlerState = KeyHandler.getInstance().keyHandlerState
+        val mappingState = keyHandlerState.mappingState
+        val vimCurrentKeySequence = mappingState.keys.toList().ifEmpty { keyHandlerState.commandBuilder.keys.toList() }
 
         // this event fires BEFORE handling by IdeaVim, so we need to construct the sequence ourselves
         val typedKeySequence = vimCurrentKeySequence + listOfNotNull(shortcut.firstKeyStroke, shortcut.secondKeyStroke)
@@ -90,9 +90,9 @@ class WhichKeyActionListener : AnActionListener {
         PopupConfig.hidePopup()
         val editor = dataContext.getData(CommonDataKeys.EDITOR) ?: return
 
-        val commandState = CommandState.getInstance(editor)
+        val keyHandlerState = KeyHandler.getInstance().keyHandlerState
 
-        val typedKeySequence = (commandState.mappingState.keys.toList().ifEmpty { commandState.commandBuilder.keys.toList() } + listOf(KeyStroke.getKeyStroke(charTyped)))
+        val typedKeySequence = (keyHandlerState.mappingState.keys.toList().ifEmpty { keyHandlerState.commandBuilder.keys.toList() } + listOf(KeyStroke.getKeyStroke(charTyped)))
             // since preceding counts are not matched for the corresponding description remove them
             .dropWhile { it.keyChar.isDigit() }
 
@@ -108,7 +108,7 @@ class WhichKeyActionListener : AnActionListener {
         val startTime = System.currentTimeMillis() // save start time for the popup delay
 
         val mappingMode = vimEditor.mode.toMappingMode()
-        val mappingState = CommandState.getInstance(editor).mappingState
+        val mappingState = KeyHandler.getInstance().keyHandlerState.mappingState
         val nestedMappings = getMappingsToDisplay(editor, typedKeySequence)
         val window = WindowManager.getInstance().getFrame(editor.project)
 
@@ -167,7 +167,7 @@ class WhichKeyActionListener : AnActionListener {
          * for example pressing "fg" should not show any nested mappings for "g" as the key press for "g" was already handled
          * and therefore any subsequent key press would not execute any of the actions presented in the popup
          */
-        if (CommandState.getInstance(editor).commandBuilder.expectedArgumentType == Argument.Type.DIGRAPH) {
+        if (KeyHandler.getInstance().keyHandlerState.commandBuilder.expectedArgumentType == Argument.Type.DIGRAPH) {
             return emptyList()
         }
 

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/WhichKeyActionListener.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/WhichKeyActionListener.kt
@@ -9,6 +9,7 @@ import com.intellij.openapi.editor.actionSystem.TypedActionHandlerEx
 import com.intellij.openapi.wm.WindowManager
 import com.maddyhome.idea.vim.VimTypedActionHandler
 import com.maddyhome.idea.vim.KeyHandler
+import com.maddyhome.idea.vim.api.VimEditor
 import com.maddyhome.idea.vim.api.injector
 import com.maddyhome.idea.vim.command.Argument
 import com.maddyhome.idea.vim.command.MappingMode
@@ -121,20 +122,21 @@ class WhichKeyActionListener : AnActionListener {
 
             // <esc> should ALWAYS close the popup without any further action
             if (typedKeySequence.size > 1 && typedKeySequence.last() == KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0)) {
-                return blockNextKeyPress(isShortcut, mappingState)
+                return blockNextKeyPress(isShortcut, vimEditor, mappingState)
             }
 
             if (!MappingConfig.processWithUnknownMapping(mappingMode, typedKeySequence)) {
-                return blockNextKeyPress(isShortcut, mappingState)
+                return blockNextKeyPress(isShortcut, vimEditor, mappingState)
             }
         } else {
             PopupConfig.showPopup(window!!, typedKeySequence, nestedMappings, startTime)
         }
     }
 
-    private fun blockNextKeyPress(isShortcut: Boolean, mappingState: MappingState) {
-        // resetting the mapping state will prevent all BUT the last key from being processed by IdeaVim
-        mappingState.resetMappingSequence()
+    private fun blockNextKeyPress(isShortcut: Boolean, editor: VimEditor, mappingState: MappingState) {
+        // resetting the key handler state will prevent all BUT the last key from being processed by IdeaVim
+        // this will reset the internal mapping state & the command builder
+        KeyHandler.getInstance().reset(editor)
 
         if (isShortcut) {
             // this is a hack to block the next action :-)

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
@@ -236,8 +236,9 @@ object MappingConfig {
 
         // is this a root key that expects a motion or digraph argument then providing this motion or digraph should not be blocked
         if (prefix.size == 1) {
+            val prefixKey = prefix.first()
             val argTypes = injector.keyGroup.getBuiltinCommandsTrie(mode).getEntries()
-                .filter { it.data !== null }
+                .filter { it.key == prefixKey && it.data !== null }
                 .map { it.data!!.instance.argumentType }
             if (argTypes.any { it == Argument.Type.MOTION || it == Argument.Type.DIGRAPH }) {
                 return true

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
@@ -3,6 +3,8 @@ package eu.theblob42.idea.whichkey.config
 import com.maddyhome.idea.vim.api.injector
 import com.maddyhome.idea.vim.command.Argument
 import com.maddyhome.idea.vim.command.MappingMode
+import com.maddyhome.idea.vim.handler.MotionActionHandler
+import com.maddyhome.idea.vim.handler.TextObjectActionHandler
 import com.maddyhome.idea.vim.key.ToKeysMappingInfo
 import com.maddyhome.idea.vim.vimscript.model.datatypes.VimString
 import eu.theblob42.idea.whichkey.model.Mapping
@@ -29,6 +31,11 @@ object MappingConfig {
      * This value only considers mappings which have more than one key stroke
      */
     private val VIM_ACTIONS = mutableMapOf<MappingMode, MutableMap<List<KeyStroke>, String>>()
+
+    /**
+     * All keystrokes that start a motion in operator-pending mode
+     */
+    private val OP_MOTION_KEYSTROKES: Set<KeyStroke>
 
     init {
         // since the VIM default mappings do not change during runtime we are extracting them once during initialization
@@ -61,6 +68,20 @@ object MappingConfig {
                     modeMap[it.first] = it.second
                 }
         }
+
+        OP_MOTION_KEYSTROKES = injector.keyGroup.getBuiltinCommandsTrie(MappingMode.OP_PENDING)
+            .getEntries()
+            // we are only interested in motions and text object actions
+            .filter { it.data?.instance is MotionActionHandler || it.data?.instance is TextObjectActionHandler }
+            .map {
+                var entry = it
+                // omit the root node as it does not contain any keystroke information
+                while (entry.parent != null && entry.parent!!.parent != null) {
+                    entry = entry.parent!!
+                }
+                entry.key
+            }
+            .toSet()
     }
 
     /**
@@ -265,9 +286,8 @@ object MappingConfig {
 
         // the prefix is a valid custom mapping and the last key is a valid motion
         // therefore we don't want to block the motion here which probably "belongs" to the mapping
-        // e.g. 'ys<motion>' (e.g. `ysiw`) and 'yss' from the builtin surround plugin
-        val operatorNode = injector.keyGroup.getBuiltinCommandsTrie(MappingMode.OP_PENDING).getEntries()
-        if (isMapping(mode, prefix) && operatorNode.any { it.key == keyStrokes.last() }) {
+        // for example 'ys<motion>' (e.g. `ysiw`) and 'yss' from the builtin surround plugin
+        if (isMapping(mode, prefix) && OP_MOTION_KEYSTROKES.contains(keyStrokes.last())) {
             return true
         }
 

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
@@ -35,12 +35,30 @@ object MappingConfig {
         for (mode in enumValues<MappingMode>()) {
             val modeMap = VIM_ACTIONS.getOrPut(mode) { mutableMapOf() }
             injector.keyGroup.getBuiltinCommandsTrie(mode).getEntries()
-                // we are only interested in VIM actions with more than one key stroke
-                .filter {
-                    it.data !== null
+                .mapNotNull {
+                    if (it.data == null) {
+                        null
+                    }
+
+                    var entry = it
+                    val tmpKeyStrokes = mutableListOf(entry.key)
+                    // rebuild the keystroke sequence by iterating through the parent nodes
+                    // omit the "root node" as it is not related to any actual keystroke
+                    while (entry.parent != null && entry.parent!!.parent != null) {
+                        entry = entry.parent!!
+                        // add parent keystrokes in front, to retain the correct order
+                        tmpKeyStrokes.add(0, entry.key)
+                    }
+
+                    // we are only interested in VIM actions with more than one keystroke
+                    if (tmpKeyStrokes.size == 1) {
+                        null
+                    }
+
+                    Pair(tmpKeyStrokes.toList(), it.data!!.actionId)
                 }
                 .forEach {
-                    modeMap[listOf(it.key)] = it.data!!.actionId
+                    modeMap[it.first] = it.second
                 }
         }
     }

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
@@ -249,7 +249,11 @@ object MappingConfig {
         // therefore we don't want to block the motion here which probably "belongs" to the mapping
         // e.g. 'ys<motion>' (e.g. `ysiw`) and 'yss' from the builtin surround plugin
         val operatorNode = injector.keyGroup.getBuiltinCommandsTrie(MappingMode.OP_PENDING).getEntries()
-        return isMapping(mode, prefix) && operatorNode.any { it.key == keyStrokes.last() }
+        if (isMapping(mode, prefix) && operatorNode.any { it.key == keyStrokes.last() }) {
+            return true
+        }
+
+        return false
     }
 
     /**

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/MappingConfig.kt
@@ -1,13 +1,8 @@
 package eu.theblob42.idea.whichkey.config
 
-import com.maddyhome.idea.vim.action.change.LazyVimCommand
 import com.maddyhome.idea.vim.api.injector
 import com.maddyhome.idea.vim.command.Argument
-import com.maddyhome.idea.vim.command.DuplicableOperatorAction
 import com.maddyhome.idea.vim.command.MappingMode
-import com.maddyhome.idea.vim.key.CommandNode
-import com.maddyhome.idea.vim.key.CommandPartNode
-import com.maddyhome.idea.vim.key.Node
 import com.maddyhome.idea.vim.key.ToKeysMappingInfo
 import com.maddyhome.idea.vim.vimscript.model.datatypes.VimString
 import eu.theblob42.idea.whichkey.model.Mapping
@@ -39,39 +34,16 @@ object MappingConfig {
         // since the VIM default mappings do not change during runtime we are extracting them once during initialization
         for (mode in enumValues<MappingMode>()) {
             val modeMap = VIM_ACTIONS.getOrPut(mode) { mutableMapOf() }
-            injector.keyGroup.getKeyRoot(mode)
+            injector.keyGroup.getBuiltinCommandsTrie(mode).getEntries()
                 // we are only interested in VIM actions with more than one key stroke
                 .filter {
-                    it.value is CommandPartNode
+                    it.data !== null
                 }
                 .forEach {
-                    extractVimActions(modeMap, listOf(it.key), it.value)
+                    modeMap[listOf(it.key)] = it.data!!.actionId
                 }
         }
     }
-
-    /**
-     * Extract the given VIM default mapping into a more usable format
-     * This function walks the tree of [Node]s and writes all found mappings along the way in to the given [vimActionsMap]
-     * The function itself does not have a return value
-     *
-     * @param vimActionsMap The (mutable) map which should be filled with the results
-     * @param keyStrokes The [KeyStroke]s related to the given [node]
-     * @param node The current mapping node
-     */
-    private fun extractVimActions(vimActionsMap: MutableMap<List<KeyStroke>, String>, keyStrokes: List<KeyStroke>, node: Node<LazyVimCommand>) {
-        if (node is CommandPartNode<LazyVimCommand>) {
-            node.map {
-                val keys = keyStrokes + listOf(it.key)
-                extractVimActions(vimActionsMap, keys, it.value)
-            }
-            return
-        }
-
-        val actionId = (node as CommandNode<LazyVimCommand>).actionHolder.instance.id
-        vimActionsMap[keyStrokes] = actionId
-    }
-
 
     /**
      * Return all nested mappings which are direct children of the given key sequence
@@ -148,12 +120,7 @@ object MappingConfig {
          */
         val keyMappingPairs = keyMapping
             .filterNotNull()
-            // check that all keystrokes are non-null (otherwise ignore for now)
-            .filter { it.filterNotNull().size == it.size }
-            .map {
-                it.map { keyStroke -> keyStroke!! }
-            }
-            // final mapping to pairs
+            // mapping to pairs
             .map {
                 Pair(it.toList(), keyMapping[it]?.getPresentableString() ?: "no description")
             }
@@ -227,10 +194,8 @@ object MappingConfig {
         // check if there is a custom user mapping for the given keystrokes
         val isCustomMapping = injector.keyGroup.getKeyMapping(mode)
             .asSequence()
-            .filterNotNull()
-            .filter { it.filterNotNull().size == it.size } // ignore mappings with null keystrokes
             .filter { it.size == keyStrokes.size }
-            .map { it.joinToString { keyStroke -> keyToString(keyStroke!!) } }
+            .map { it.joinToString { keyStroke -> keyToString(keyStroke) } }
             .find { it == seq } != null
 
         if (isCustomMapping) {
@@ -271,10 +236,9 @@ object MappingConfig {
 
         // is this a root key that expects a motion or digraph argument then providing this motion or digraph should not be blocked
         if (prefix.size == 1) {
-            val prefixKey = prefix.first()
-            val argTypes = injector.keyGroup.getKeyRoot(mode)
-                .filter { it.key == prefixKey && it.value is CommandNode }
-                .map { (it.value as CommandNode).actionHolder.instance.argumentType }
+            val argTypes = injector.keyGroup.getBuiltinCommandsTrie(mode).getEntries()
+                .filter { it.data !== null }
+                .map { it.data!!.instance.argumentType }
             if (argTypes.any { it == Argument.Type.MOTION || it == Argument.Type.DIGRAPH }) {
                 return true
             }
@@ -283,12 +247,8 @@ object MappingConfig {
         // the prefix is a valid custom mapping and the last key is a valid motion
         // therefore we don't want to block the motion here which probably "belongs" to the mapping
         // e.g. 'ys<motion>' (e.g. `ysiw`) and 'yss' from the builtin surround plugin
-        val operatorNode = injector.keyGroup.getKeyRoot(MappingMode.OP_PENDING)
-        if (isMapping(mode, prefix) && operatorNode.any { it.key == keyStrokes.last() }) {
-            return true
-        }
-
-        return false
+        val operatorNode = injector.keyGroup.getBuiltinCommandsTrie(MappingMode.OP_PENDING).getEntries()
+        return isMapping(mode, prefix) && operatorNode.any { it.key == keyStrokes.last() }
     }
 
     /**

--- a/src/main/kotlin/eu/theblob42/idea/whichkey/config/PopupConfig.kt
+++ b/src/main/kotlin/eu/theblob42/idea/whichkey/config/PopupConfig.kt
@@ -70,6 +70,7 @@ object PopupConfig {
      * @param nestedMappings A [List] of nested mappings to display
      * @param startTime Timestamp to consider for the calculation of the popup delay
      */
+    @OptIn(DelicateCoroutinesApi::class)
     fun showPopup(ideFrame: JFrame, typedKeys: List<KeyStroke>, nestedMappings: List<Pair<String, Mapping>>, startTime: Long) {
         if (nestedMappings.isEmpty()) {
             return

